torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -161,6 +161,13 @@ def __init__(
         self._init_extensions()
         self.all_gather_outputs: List[torch.Tensor] = []
         self._param_fqn: Optional[str] = None  # prefixed from root module
+        # TODO: Remove this padding logic once DTensor pads the local tensor:
+        # https://github.com/pytorch/pytorch/issues/113045
+        self._post_load_hook_handle = (
+            module_info.module.register_load_state_dict_post_hook(
+                lambda *args, **kwargs: self.reset_sharded_param()
+            )
+        )
```

Comment:
This is needed to support `load_state_dict(assign=True)` today.

In the future, we want (1) `DTensor` to pad its local tensor and (2) for users to use the `swap_tensors` path for `load_state_dict()` with FSDP2. Together these should allow us to remove this hook.